<html>
<head>
<title>crux create-index</title>
<link href="./crux.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<H1 ALIGN=CENTER>crux create-index</H1>

<B>Usage:</B>

<blockquote>
<code>crux create-index [options]
  &lt;protein&nbsp;input&nbsp;file&gt;
  &lt;index&nbsp;name&gt;
</code>
</blockquote>

<B>Description:</B>

<blockquote>
Given a protein fasta sequence database as input, generate an index of
all of its peptides and save it to disk.  Optionally, the index may
also include decoy peptides to be used for post-processing commands
such as <code>crux percolator</code>, <code>crux q-ranker</code>
<code>crux compute-q-values</code> or <code>crux barista</code>.  The
index can be provided as input to <code>crux search-for-matches</code>
instead of the fasta file to speed up the search.</blockquote>

<p>
<B>Input:</B>

<UL>
  <LI>&lt;protein input file&gt; &ndash;
  The name of the file (in fasta format) from which to parse proteins.
  <LI>&lt;index name&gt; &ndash;
  The name of the directory where the newly created index will be placed.
	
</UL>
<B>Output:</B>

<UL>

<p>
The program writes four types of files to the specified output
directory: binary index files, a binary sequence file, and a text file
with settings, a fasta file with decoy protein sequences to be used
during search.  The binary sequence file will have the same name as
the input fasta file, but ending with '<code>binary_fasta</code>'. The
fasta file with decoy protein sequences will have the same name is the
input fasta file, with ending '<code>-random.fasta</code> if the
random shuffling is used and '<code>-reverse.fasta</code> if the
protein sequences are reversed (see options below).  The text file is
named '<code>README</code>'.</p>

</UL>
<B>Options:</B>

<UL>
	
<LI>
<code>--min-mass &lt;float&gt;</code> &ndash;
The minimum neutral mass of the peptides to place in the
index. Default = 200.
	
<LI>
<code>--max-mass &lt;float&gt;</code> &ndash;
The maximum neutral mass of the peptides to place in index. Default =
7200.
<LI>
<code>--min-length &lt;int&gt;</code> &ndash;
The minimum length of the peptides to place in the index. Default = 6.
	
<LI>
<code>--max-length &lt;int&gt;</code> &ndash;
The maximum length of the peptides to place in the index. Default = 50.

<LI>
<code>--enzyme trypsin|trypsin/p|chymotrypsin|elastase|clostripain|cyanogen-bromide|idosobenzoate|proline-endopeptidase|staph-protease|asp-n|lys-c|lys-n|arg-c|glu-c|pepsin-a|elastase-trypsin-chymotrypsin|no-enzyme</code> &ndash;
Enzyme to use for <i>in silico</i> digestion of protein
sequences. Used in conjunction with the <code>digestion</code>
and <code>missed-cleavages</code> options. Use 'no-enzyme' for
non-specific digestion.  Digestion rules are as follows: enzyme name
[cuts after one of these residues]|{but not before one of these
residues}. trypsin [RK]|{P}, trypsin/p [RK]|[], elastase [ALIV]|{P},
chymotrypsin [FWYL]|{P}, clostripain [R]|[], cyanogen-bromide [M]|[],
iodosobenzoate [W]|[], proline-endopeptidase [P]|[], staph-protease [E]|[],
elastase-trypsin-chymotrypsin [ALIVKRWFY]|{P}, asp-n []|[D], lys-c [K]|{P},
lys-n []|[K], arg-c [R]|{P}, glu-c [DE]|{P}, pepsin-a [FL]|{P}.
Default = trypsin.

<li>
<code>--custom-enzyme &lt;residues before cleavage&gt;|&lt;residues after cleavage&gt;</code> &ndash;
Specify rules for <i>in silico</i> digestion of protein sequences.
Overrides the <code>enzyme</code> option.  Two lists of residues are
given enclosed in square brackets or curly braces and separated by a
|. The first list contains residues required/prohibited before the
cleavage site and the second list is residues after the cleavage site.
If the residues are required for digestion, they are in square
brackets, '[' and ']'.  If the residues prevent digestion, then they
are enclosed in curly braces, '{' and '}'.  Use X to indicate all
residues.  For example, trypsin cuts after R or K but not before P
which is represented as <code>[RK]|{P}</code>.  AspN cuts after any
residue but only before D which is represented as <code>[X]|[D]</code>.

<LI>
<code>--digestion full-digest|partial-digest </code>&ndash;
Degree of digestion used to generate peptides . Either both ends
(full-digest) or at least one end (partial-digest) of a peptide must
conform to enzyme specificity rules. Used in conjunction with
the <code>enzyme</code> or <code>custom-enzyme</code> option
when <code>enzyme</code> is not set to to 'no-enzyme'.   Default
full-digest.

<LI>
<code>--missed-cleavages &lt;int&gt;</code> &ndash; Include in the
index peptides containing up to &lt;int&gt; missed cleavage sites.
Default = 0.</li>

<LI>
<code>--isotopic-mass average|mono</code> &ndash;
Specify the type of isotopic masses to use when calculating the
peptide mass. Default = average.</li>

<li>
<code>--peptide-list T|F</code> &ndash; Create in the output
directory a text file listing all of the peptides in the database,
along with their neutral masses, one per line.  Default = F.</li>

<li>
<code>--decoys &lt;string&gt;</code> &ndash; Include a decoy version
of every peptide by shuffling or reversing the target sequence.
Possible values are <code>none, reverse, protein-shuffle,
peptide-shuffle</code>. Use 'reverse' to reverse each protein sequence,
'protein-shuffle' to shuffle each protein sequence, or 'peptide-shuffle'
to shuffle the sequence between enzyme cleavage sites, leaving the
termini in place. Use 'none' for no decoys.  Default=peptide-shuffle. 

<li>
<code>--overwrite T|F</code> Replace existing files if true (T) or
fail when trying to overwrite a file if false (F). Default = F.

<LI>
<code>--parameter-file &lt;string&gt;</code>
A file containing command-line or additional parameters.  See the <a
href="parameter-file.html">parameter documentation</a> page for
details.  Default = no parameter file.

<LI>
<code>--verbosity &lt;0-100&gt;</code> &ndash; 
Specify the verbosity of the current processes. Each level prints the
following messages including all those at lower verbosity levels:
0-fatal errors, 10-non-fatal errors, 20-warnings, 30-information on
the progress of execution, 40-more progress information, 50-debug
info, 60-detailed debug info. Default = 30.

</ul>

<B>Parameter file options:</B>
<ul>

<li>
<code>&lt;A-Z&gt &lt;float&gt;</code> &ndash;
Specify static modifications.  This is a mass change applied to the
given amino acid (in single-letter-code A thru Z) for every peptide in
which it occurs.  Use the <code>mod</code> option when searching to
generate peptides both with and without the mass change.  Default C=57.
</ul>

<b>Increasing the file limit:</b>

<ul>

<p>
The Crux indexing scheme relies on having a large number of files open
simultaneously.  On some operating systems, this will result in an
error message like this:</p>

<pre>
     WARNING: cannot open all file handlers needed
</pre>

<p>
On a Unix system, you can find out the maximum number of files that
can be open simultaneously using a command like this:</p>

<pre>
     ulimit -a
</pre>

<p>
You can then use</p>

<pre>
     ulimit -n 1024
</pre>

<p>
to reset the limit (in this case, to 1024 files). Note that once you
have set the limit on open files handles in a shell you cannot
increase it again for the duration of that shell.  However, if you
start a new shell, you can once again increase the open file handle
limit. New shells will remember the last limit set.</p>

</ul>

<b>Known Bugs</b>
<p>
In cygwin, the <code>--overwrite T</code> option does not work.  A 
temporary directory is created with the new index but fails to be
re-named correctly.  As a work-around, delete the existing index and
rename the new temporary index with the correct name.</p>

<hr></hr>

<p>
<a href="crux.html">Crux home</a></p>

</blockquote>
</body>
</html>
