<html>
<head>
<title>Crux tandem mass spectrometry analysis software</title>
<link href="./crux.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<!--
  N.B. I created this image in PowerPoint, exported to PDF and then
  created the PNG using the following command:

  convert -density 300 -resize 1125x1125 schematic.pdf schematic.png

  WSN 6 May 2014
-->

<!--
 Updated map
 SJM 8 May 2014
-->

<center>
<img  src="schematic.png" usemap="#commands-files">
<map name="commands-files">
  <area shape="rect" coords="253,135,349,169" alt="tide-index" href="tide-index.html"/>
  <area shape="rect" coords="260,247,342,278" alt="bullseye" href="bullseye.html"/>
  <area shape="rect" coords="510,100,612,132" alt="tide-search" href="tide-search.html"/>
  <area shape="rect" coords="528,191,595,222" alt="comet" href="comet.html"/>
  <area shape="rect" coords="512,281,613,335" alt="search-for-xlinks" href="crux-search-for-xlinks.html"/>
  <area shape="rect" coords="776,100,873,132" alt="percolator" href="percolator.html"/>
  <area shape="rect" coords="791,160,861,192" alt="barista" href="barista.html"/>
  <area shape="rect" coords="783,222,866,252" alt="q-ranker" href="q-ranker.html"/>
  <area shape="rect" coords="784,281,867,334" alt="spectral-counts" href="spectral-counts.html"/>
</map>

</center>

<p>
Crux is a software toolkit for tandem mass spectrometry analysis.  If
you use Crux in your research, please cite</p>

<blockquote>
Christopher Y. Park, Aaron A. Klammer, Lukas K&auml;ll, Michael
J. MacCoss and William Stafford
Noble.  <a href="http://noble.gs.washington.edu/papers/park2008rapid.html">"Rapid
and accurate peptide identification from tandem mass spectra."</a>
<i>Journal of Proteome Research.</i> 7(7):3022-3027, 2008.</li>
</blockquote>

<p>
For a more up-to-date description of Crux, please read</p>

<blockquote>
Sean McIlwain, Kaipo Tamura, Attila Kertesz-Farkas,
Charles E.  Grant, Benjamin Diament, Barbara Frewen, J. Jeffry
Howbert, Michael R. Hoopmann, Lukas K&auml;ll, Jimmy K. Eng, Michael
J. MacCoss and William Stafford
Noble.  <a href="http://pubs.acs.org/doi/abs/10.1021/pr500741y">"Crux:
rapid open source protein tandem mass spectrometry
analysis."</a> <i>Journal of Proteome Research.</i>
13(10):4488-4491, 2014.
</blockquote>


<p />
<a href="download" title="Download Crux Toolkit" class="sfdl">
    <span>
        <b>Download Crux Toolkit</b>
    </span>
</a>
<div style="clear:both;"></div>
<p/>
Crux is provided as a single executable that supports the following
commands:</p>

<table>

<tr>
<td colspan="2"><b>Primary commands</b></td>
</tr>

<tr>
<td><a href="bullseye.html">bullseye</a></td>
<td>Assign high resolution precursor m/z values to MS/MS data using the Hardkl&ouml;r algorithm.
</tr>

<tr>
<td>
<a href="tide-index.html">tide-index</a></td>
<td>
Create an index of all peptides in a fasta file, for use in subsequent
calls to <code>tide-search</code>.</td>
</tr>

<tr>
<td>
<a href="tide-search.html">tide-search</a></td>
<td>
Search a collection of spectra against a sequence database, returning
a collection of peptide-spectrum matches (PSMs).  This is a fast
search engine but requires that you first build an index
with <code>tide-index</code>.</td>
</tr>

<tr>
<td>
<a href="comet.html">comet</a></td>
<td>
Search a collection of spectra against a sequence database, returning
a collection of PSMs.  This search engine runs directly on a protein
database in FASTA format.</td>
</tr>

<tr>
<td>
<a href="percolator.html">percolator</a></td>
<td>
Re-rank a collection of PSMs using the Percolator algorithm.
Optionally, also produce protein rankings using the Fido algorithm.</td></tr>

<tr>
<td>
<a href="q-ranker.html">q-ranker</a></td>
<td>
Re-rank a collection of PSMs using the Q-ranker algorithm.</td></tr>

<tr>
<td>
<a href="barista.html">barista</a></td>
<td>
Rank PSMs, peptides and proteins, assigning a confidence measure to each
identification.</td></tr> 

<tr>
<td>
<a href="crux-search-for-xlinks.html">search-for-xlinks</a></td>
<td>
Search a collection of spectra against a sequence database, finding
cross-linked peptide matches.</td></tr>

<tr>
<td>
<a href="spectral-counts.html">spectral-counts</a></td>
<td>
Quantify peptides or proteins using one of three spectral counting methods.</td></tr>

<tr>
<td colspan="2"><b>Utilities</b></td>
</tr>

<tr>
<td>
<a href="calibrate-scores.html">calibrate-scores</a></td>
<td>
Assign two types of statistical confidence measures (q-values and
posterior error probabilities) to each PSM in a given set.</td></tr>

<tr>
<td><a href="make-pin.html">make-pin</a></td>
<td>Given a set of search results files, generate a pin file for input to
<code>crux percolator</code></td>
</tr>

<tr>
<td>
<a href="crux-predict-peptide-ions.html">predict-peptide-ions</a></td>
<td>
Given a peptide and a charge state, predict the m/z values of the
resulting fragment ions.</td>
</tr>

<tr>
<td><a href="hardklor.html">hardkl&ouml;r</a></td>
<td>Identify isotopic distributions from high-resolution mass spectra.</td>
</tr>

<tr>
<td>
<a href="crux-print-processed-spectra.html">print-processed-spectra</a></td>
<td>
Process spectra as for scoring xcorr and print the results to a file.</td></tr>

<tr>
<td>
<a href="generate-peptides.html">generate-peptides</a></td>
<td>
Extract from a given set of protein sequences a list of target and
decoy peptides fitting the specified criteria.</td>
</tr>

<tr>
<td>
<a href="crux-get-ms2-spectrum.html">get-ms2-spectrum</a></td>
<td>
Extract one or more fragmentation spectra, specified by scan number, from
an <code>MS2</code> file.</td>
</tr>

<tr>
<td>
version</td>
<td>
Print the Crux version number to standard output, then exit.</td>
</tr>

<tr>
<td colspan="2"><b>Utilities for processing tab-delimited text files</b>
</tr>

<tr>
<td>
<a href="extract-columns.html">extract-columns</a></td>
<td>Print specified columns from a tab-delimited file.</td>
</tr>

<tr>
<td>
<a href="extract-rows.html">extract-rows</a></td>
<td>Print specified rows from a tab-delimited file.</td>
</tr>

<tr>
<td>
<a href="stat-column.html">stat-column</a></td>
<td>Collect summary statistics from a column in a tab-delimited file.</td>
</tr>

<tr>
<td>
<a href="sort-by-column.html">sort-by-column</a></td>
<td>Sort a tab-delimited file by a column.</td>
</tr>

</table>

<h2>File formats</h2>

<table id="fileFormatsTable" border="1">

<tr>
<td>&nbsp;</td>
<td>bullseye</td>
<td>tide-index</td>
<td>tide-search</td>
<td>comet</td>
<td>percolator</td>
<td>q-ranker</td>
<td>barista</td>
<td>search-for-xlinks</td>
<td>spectral-counts</td>
</tr>

<tr>
<td>
<a href="ms2-format.html">MS1</a>
</td>
<td>in</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>
<a href="ms2-format.html">MS2</a>
(<a href="data/demo.ms2">sample</a>) / CMS2,
<a href="http://www.matrixscience.com/help/data_file_help.html">MGF</a>
</td>
<td>in/out</td>
<td>&nbsp;</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>in</td>
</tr>

<tr>
<td>
<a href="http://www.psidev.info/mzml_1_0_0">mzML</a>,
<a href="http://tools.proteomecenter.org/wiki/index.php?title=Formats:mzXML">mzXML</a>
</td>
<td>in</td>
<td>&nbsp;</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>in</td>
</tr>

<tr>
<td>
<a href="http://sjsupport.thermofinnigan.com/public/detail.asp?id=703">Thermo .raw files</a>
(Windows only)
</td>
<td>in</td>
<td>&nbsp;</td>
<td>in</td>
<td>in</td>
<td>&nbsp;</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>in</td>
</tr>

<tr>
<td>
<a href="http://proteowizard.sourceforge.net/formats.shtml">Other
  proprietary vendor formats</a>  (Windows only)
</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>in</td>
</tr>

<tr>
<td>
<a href="fasta-format.html">FASTA</a>
(<a href="sample.fasta">sample</a>)
</td>
<td>in</td>
<td>in</td>
<td>&nbsp;</td>
<td>in</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>
database index
</td>
<td>&nbsp;</td>
<td>out</td>
<td>in</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>
<a href="txt-format.html">Tab-delimited text</a>
(<a href="data/search.target.txt">sample</a>)
</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>out</td>
<td>out</td>
<td>in/out</td>
<td>in/out</td>
<td>in/out</td>
<td>out</td>
<td>in/out</td>
</tr>

<tr>
<td>
<a href="http://tools.proteomecenter.org/wiki/index.php?title=Formats:pepXML">
pepXML</a>
</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>out</td>
<td>out</td>
<td>in/out</td>
<td>out</td>
<td>out</td>
<td>&nbsp;</td>
<td>in</td>
</tr>

<tr>
<td>
<a href="https://github.com/percolator/percolator/wiki/Interface">PIN</a>
</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>out</td>
<td>&nbsp;</td>
<td>in</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>
<a href="https://github.com/percolator/percolator/wiki/Interface">POUT</a>
</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>
<a href="http://www.psidev.info/mzidentml">mzIdentML</a>
</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>out</td>
<td>out</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in</td>
</tr>

<tr>
<td>
<a href="sqt-format.html">SQT</a>
</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>out</td>
<td>out</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>&nbsp;</td>
<td>in</td>
</tr>

<tr>
<td>
<a href="features.html">Features</a>
</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>
<a href="barista-xml-format.html">Barista XML</a>
</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>

<tr>
<td>
<a href="parameter-file.html">Crux parameters</a>
(<a href="default.params">sample</a>)
</td>
<td>in/out</td>
<td>in/out</td>
<td>in/out</td>
<td>in/out</td>
<td>in/out</td>
<td>in/out</td>
<td>in/out</td>
<td>in/out</td>
<td>in/out</td>
</tr>

</table>

<h2>Tutorials</h2>

<ul>

<li>
<a href="install-tutorial.html">Installation</a></li>

<li>
<a href="start-tutorial.html">Getting started with Crux</a></li>

<li>
<a href="search-tutorial.html">Running a simple search using Tide and
Percolator</a></li>

<li>
<a href="customize-tutorial.html">Customization and search options</a></li>

<li>
<a href="spectral-counts-tutorial.html">Using spectral-counts</a></li>

</ul>


<h2>Important links</h2>

<ul>
 
<li>
<a href="crux-faq.html">Crux frequently asked questions</a></li>

<li>
<a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Crux
Google Scholar profile</a></li>

<li>
To receive announcements of new versions, sign up for the Crux users 
<a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users">mailing
list.</a></li>

<li>
For support, and to discuss Crux with other users and the developers,
sign up for the Crux users
<a href="http://groups.google.com/group/crux-users">Google
  group.</a></li>

<li>
Follow the Crux project
on <a href="https://sourceforge.net/projects/cruxtoolkit/">Sourceforge</a>.
The "Issues" list on Sourceforge includes known bugs as well as
planned new features.</li>

<li>
Access the most recent source code from the Crux main development
trunk using the following Subversion command:<br>
<code>&nbsp;&nbsp;&nbsp;svn checkout svn://svn.code.sf.net/p/cruxtoolkit/code/crux/trunk crux-trunk</code></li>

<li>
<a href="release-notes.html">Release notes.</a></li>

<li>
Crux is covered by an <a href="http://www.apache.org/licenses/LICENSE-2.0">Apache
license</a>.</li>

</ul>

<hr></hr>

<center>
<p>
The original version of Crux was written by Chris Park and Aaron Klammer
under the supervision
of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
MacCoss</a>
and <a href="http://noble.gs.washington.edu/~noble">Prof. William
Stafford Noble</a> in the Department of Genome Sciences at the
University of Washington, Seattle.  The complete list of contributors
can be found <a href="contributors.html">here</a>.</p>

<p>Maintenance and development of Crux is funded by the National
Institute of General Medical Sciences grants R01&nbsp;GM096306
and P41&nbsp;GM103533.</p>
</center>

<hr></hr>

<p>
Please send comments and questions to cegrant@uw.edu.</p>

</blockquote>
</body>
</html>
