<html>
<head>
<title>crux sequest-search</title>
<link href="./crux.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<H1 ALIGN=CENTER>crux sequest-search</H1>
<!-- open source warning
<h2>
You are using the open source version of Crux.
Due to intellectual property issues,
we are unable to provide database search functionality in this version.
To obtain a license for the fully functional version of Crux that includes
the database search tools please visit 
<a href="http://depts.washington.edu/ventures/UW_Technology/Express_Licenses/crux.php">UW TECHTRANSFER</a>.
</h2>
open source warning -->
<B>Usage:</B>

<blockquote>
<code>crux sequest-search [options]
&lt;ms2&nbsp;input&nbsp;filename&gt; &lt;protein&nbsp;input&gt;
</code>
</blockquote>

<B>Description:</B>

<blockquote>
<p>
This command searches a protein database with a set of spectra, using
an algorithm similar to that employed by the SEQUEST database search
software.  This command differs from the <code>crux
search-for-matches</code> command in the following respects:</p>
<ul>
<li>
Candidate peptides (i.e., peptides that lie within the user-specified
precursor window) are first ranked with the SEQUEST Sp score.  The top
500 matches are then re-scored using XCorr.</li>
<li>
The program will not compute p-values using the Weibull empirical
curve fitting procedure.</li>
<li>
By default, the theoretical spectrum used to compute XCorr includes
two flanking peaks on either side of each b- and y-ion.  These can be
turned off using the <code>use-flanking-peaks</code> parameter.</li>
<li>
The program produces <a href="sqt-format.html">SQT</a> format output
files.</li>
</ul>

<p>
Similar to <code>crux search-for-matches</code>, the input protein
database may either be in FASTA format or it may be a binary index
created by <code>crux create-index</code>.  Using an index will
typically yield much faster search speeds.</p>

<p>
This command handles modifications in the same way
as <a href="crux-search-for-matches.html"><code>crux
search-for-matches</code></a>.</p>

</blockquote>

<p>
<B>Input:</B>

<UL>

<LI>
&lt;ms2 &gt; &ndash; The name of the file (in MS2 format) from which
  to parse the spectra. </li>

<LI>
&lt;protein input&gt; &ndash; 
  The name of the file in fasta format or the directory containing a
  protein index from which to retrieve proteins and peptides.</li>

</UL>

<B>Output:</B>

<p>
The program writes files to the folder <code>crux-output</code> by
default.  The name of the output folder can be set by the user using
the <code>--output-dir</code> option.  The following files will be
created:</p>
<ul>

<li>
<b>sequest.params.txt</b>: a file containing the name and value of all
parameters/options for the current operation.  Not all parameters in
the file may have been used in the operation.  The resulting file can
be used with the <code>--parameter-file</code> option for other crux
programs.
</li>

<li>
<b>sequest.target.sqt</b>: an <a href="sqt-format.html">SQT</a> file
containing the PSMs.
</li>

<li>
<b>sequest.target.txt</b>: a tab-delimited text file containing the
PSMs.  See <a href="txt-format.html">txt file format</a> for a list of
the fields.  This and the sequest.decoy.txt file(s) can be used as
input for <code>crux</code> post-search functions such as <code>
q-ranker</code>.
</li>

<li>
<b>sequest.log.txt</b>: a log file containing a copy of all messages
that were printed to stderr.
</li>

</ul>

<p>
If decoys are enabled using <code>--num-decoys-per-target</code>, then
search.decoy.sqt and search.decoy.txt are also produced.</p>

<B>Options:</B>
<UL>

<li>
<code>--fileroot &lt;string&gt;</code> &ndash; 
The <code>fileroot</code> string will be added as a prefix to all
output file names. Default = none.</li>

<li>
<code>--output-dir &lt;filename&gt;</code> &ndash; 
The name of the directory where output files will be created.
Default = crux-out.</li>


<li>
<code>--overwrite &lt;T|F&gt;</code> &ndash;
Replace existing files if true (T) or fail when trying to overwrite a
file if false (F). Default =  F.</li>

<li>
<code>--decoys &lt;string&gt;</code> &ndash; Include a decoy version
of every peptide by shuffling or reversing the target sequence.
Possible values are <code>none, reverse, protein-shuffle,
peptide-shuffle</code>. Use 'reverse' to reverse each protein sequence,
'protein-shuffle' to shuffle each protein sequence, or 'peptide-shuffle'
to shuffle the sequence between enzyme cleavage sites, leaving the
termini in place. Use 'none' for no decoys.  Default=peptide-shuffle. 
</li>

<LI>
<code>--num-decoys-per-target &lt;n&gt;</code> &ndash; 
Specify the number of decoy peptides to search for every target
peptide searched. Control where the decoys are returned (to what
files) with <code>--decoy-location</code>. At least one decoy set (in
its own file) is required to run the algorithm 'percolator' in a
subsequent crux run.  Default = 2. </li>

<li>
<code>--decoy-location &lt;target-file | one-decoy-file |
separate-decoy-files&gt;</code> &ndash; 
File(s) in which decoy results are returned.  Only applies
when <code>num-decoys-per-target</code> is not zero.  Use
'target-file' to mix target and decoy psms in one file. Use
'one-decoy-file' to print target psms to one file and all decoys to a
separate file.  Use 'separate-decoy-files' to print one .txt file for
each decoy set.  (<code>crux percolator</code> accepts up to two
search.decoy.txt files. <code>crux q-ranker</code> accepts only one
search.decoy.txt file.) Default = separate-decoy-files.</li>

<li>
<code>--spectrum-parser pwiz|mstoolkit</code> &ndash;
Specify the parser to use for reading in MS/MS spectra. The default,
ProteoWizard parser should be able to read the MS/MS file formats listed
<a href="http://proteowizard.sourceforge.net/formats.shtml">here</a>. The
alternative is <a href="mstoolkit.html">MSToolkit parser</a>. If the
ProteoWizard parser fails to read your files properly, you may want to try the
MSToolkit parser instead. Default = pwiz.</li>

<LI>
<code>--spectrum-min-mz &lt;float&gt;</code> &ndash; 
The lowest spectrum m/z to search in the ms2 file. Default = 0.0
</li>

<LI>
<code>--spectrum-max-mz &lt;float&gt;</code> &ndash; 
The highest spectrum m/z to search in the ms2 file. Default = no maximum.
</li>

<LI> 
<code>--spectrum-charge &lt;1|2|3|all&gt;</code> &ndash; 
The spectrum charges to search. With 'all' every spectrum will be
searched and spectra with multiple charge states will be searched once
at each charge state.  With 1, 2, or 3 only spectra with that charge
will be searched. Default = all. 
</li>

<LI>
<code>--scan-number &lt;number range&gt;</code> &ndash;
A single scan number or a range of numbers to be searched. Range
should be specified as 'first-last' which will include scans 'first'
and 'last'.  Default = search all spectra.
</li>

<LI><code>--parameter-file &lt;filename&gt;</code> &ndash; 
A file containing command-line or additional parameters.  See
the <a href="parameter-file.html">parameter documentation</a> page
for details.  Default = no parameter file.
</li>

<LI>
<code>--verbosity &lt;0-100&gt;</code> &ndash; 
Specify the verbosity of the current processes. Each level prints the
following messages, including all those at lower verbosity levels:
0-fatal errors, 10-non-fatal errors, 20-warnings, 30-information on
the progress of execution, 40-more progress information, 50-debug
info, 60-detailed debug info. Default = 30.
</li>

</UL>

<B>Parameter file options:</B>

<ul>

<LI>
<code>--min-peaks &lt;int&gt;</code> &ndash;
The minimum number of peaks a spectrum must have in order to be
searched.  Default = 20.
</li>

<li>
<code>fragment-mass &lt;average|mono&gt;</code> &ndash; 
Which isotopes to use in calculating fragment ion mass (average,
mono). Default = mono.</li>

<li>
<code>use-flanking-peaks &lt;T|F&gt;</code> &ndash;
Turn on or off the peaks flanking the b/y ions.  For <code>crux 
search-for-matches</code>, default = F; for <code>crux
  sequest-search</code>, default = T; for <code> crux
search-for-xlinks</code>, default = T.</li>
 
<li>
<code>precursor-window &lt;float&gt;</code> &ndash; Tolerance used for
matching peptides to spectra.  Peptides must be within +/-
'<code>precursor-window</code>' of the spectrum value.  Definition of
precursor window units depends
upon <code>precursor-window-type</code>. Default = 3.0.</li>

<li>
<code>precursor-window-type &lt;mass|mz|ppm&gt;</code> &ndash; Specify
the units for the window that is used to select peptides around the
precursor mass location (mass, mz, ppm).  The magnitude of the window
is defined by the <code>precursor-window</code> option, and candidate
peptides must fall within this window.  For the <code>mass</code>
window-type, the spectrum precursor singly charged mass (m+h) is
converted to mass, and the window is defined as that mass
&plusmn; <code>precursor-window</code>. If the m+h value is not
available, then the mass is calculated from the precursor
mass-to-charge (m/z) and provided charge.  The peptide mass is
computed as the sum of the average amino acid masses plus 18 Da for
the terminal OH group.  The <code>mz</code> window-type calculates the
window as spectrum precursor m/z
&plusmn; <code>precursor-window</code> and then converts the resulting
m/z range to the peptide mass range using the precursor charge.  For
the parts-per-million (<code>ppm</code>) window-type, the spectrum
mass is calculated as in the <code>mass</code> type.  The lower bound
of the mass window is then defined as the spectrum mass / (1.0 +
(<code>precursor-window</code> / 1000000)) and the upper bound is
defined as spectrum mass / (1.0 - (<code>precursor-window</code> /
1000000)).  Default = mass.</li>

<li>
<code>top-match &lt;int&gt;</code> &ndash; 
The number of psms per spectrum written to the output files. Default =
5.
</li>

<li>
<code>max-rank-preliminary &lt;int&gt;</code> &ndash;
The number of PSMs to score with XCorr after ranking PSMs by Sp.
Default = 500.
</li>

<li>

<code>mod &lt;mass&nbsp;change&gt;:&lt;aa&nbsp;list&gt;:&lt;max&nbsp;per&nbsp;peptide&gt;:&lt;prevents&nbsp;cleavage&gt;:&lt;prevents&nbsp;cross-link&gt;
</code> &ndash; 
Consider modifications on any amino acid in aa list with at most
max-per-peptide in one peptide.  This parameter may be included
with different values multiple times so long as the total number of
<code>mod</code>, <code>cmod</code>, and <code>nmod</code> parameters
does not exceed 11.  The <code>prevents&nbsp;cleavage</code> and 
<code>prevents&nbsp;cross-link</code> are T/F optional arguments for 
describing whether the modification prevent enzymatic cleavage or 
cross-linking respectively. The same modifications must be given for any
post-search process (<code>crux compute-q-values</code>, <code>crux
q-ranker</code>, <code>crux percolator</code>). Default = no
variable modifications.
</li>

<li>
<code>cmod &lt;mass change&gt;:&lt;max distance from protein
C-terminus&gt;</code> &ndash;
Consider modifications on the C-terminus of any peptide whose
C-terminus is no more than max-distance residues from the protein
C-terminus.  Use -1 to consider the C-terminus of all peptides
regardless of position in the protein.  This parameter may be included
with different values multiple times so long as the total number
of <code>mod</code>, <code>cmod</code>, and <code>nmod</code>
parameters does not exceed 11. The same modifications must be given
for any post-search process (<code>crux
  compute-q-values</code>, <code>crux q-ranker</code>, <code>crux
  percolator</code>). Default = no c-terminal modifications.
</li>

<li>
<code>nmod &lt;mass change&gt;:&lt;max distance from protein N-terminus&gt;
&ndash;</code> 
Consider modifications on the N-terminus of any peptide whose
N-terminus is no more than max-distance residues from the protein
N-terminus.  Use -1 to consider the N-terminus of all peptides regardless
of position in the protein. This parameter may be included with
different values multiple times so long as the total number of
<code>mod</code>, <code>cmod</code>, and <code>nmod</code> parameters
does not exceed 11. The same modifications must be given for any
post-search process (<code>crux compute-q-values</code>, <code>crux 
q-ranker</code>, <code>crux percolator</code>). Default = no
n-terminal modifications.
</li>

<li>
<code>cmod-fixed &lt;mass change&gt&ndash;</code>
Add a modification of the given mass change to the C-terminus of every
peptide.
</li>

<li>
<code>nmod-fixed &lt;mass change&gt&ndash;</code>
Add a modification of the given mass change to the N-terminus of every
peptide.
</li>

<li>
<code>max-mods &lt;int&gt;</code> &ndash;
The maximum number of modifications that can be applied to a single
peptide. Default = no limit.
</li>

<li>
<code>max-aas-modified &lt;int&gt;</code> &ndash;
The maximum number of modified amino acids that can appear in one
peptide.  Each aa can be modified multiple times. Default = no limit.
</li>

<li>
<code>mod-mass-format &lt;mod-only|total|separate&gt;<code> &ndash;
Specify how sequence modifications are reported in various ouptut
files.  Each modification is reported as a number enclosed in square
braces following the modified reside; however, the number may
correspond to one of three different masses: (1) 'mod-only' reports
the value of the mass shift induced by the modification; (2) 'total'
reports the mass of the residue with the modification (residue mass
plus modification mass); (3) 'separate' is the same as 'mod-only', but
multiple modifications to a single amino acid are reported as a
comma-separated list of values. For example, suppose amino
acid <i>D</i> has an unmodified mass of 115 as well as two
modifications of masses +14 and +2.  In this case, the amino acid
would be reported as D[16] with 'mod-only', D[131] with 'total', and
D[14,2] with 'separate'.</li>

<li>
<code>precision &lt;int&gt;</code> &ndash;
Set the precision (number of significant digits) for scores
written to text files. Default = 8.
</li>

<li>
<code>print-search-progress &lt;int&gt;</code> &ndash;
Show search progress by printing every n spectra searched. Set to
0 to show no search progress.  Default = 10.
</li>

<li>
<code>reverse-sequence &lt;T|F&gt;</code> &ndash;
Generate decoy sequences by reversing the peptide rather than by
shuffling.  The first and last residues of the sequence are not
changed.  If the target sequence is a palindrome (the same when
reversed), then the decoy will be generated by shuffling and a note to
that effect will be printed at verbosity level 40 (DETAILED
INFO). Default = generate decoys by shuffling.
</li>

<p>
<i>NOTE: the following parameters are also used when creating an index
and must be compatible with any index used.</i></p>

<LI>
<code>min-mass &lt;float&gt;</code> &ndash; 
The minimum neutral mass of the peptides to place in the index. Default
= 200.
</li>

<LI>
<code>max-mass &lt;float&gt;</code> &ndash; 
The maximum neutral mass of the peptides to place in index. Default = 7200.
</li>
	
<LI>
<code>min-length &lt;int&gt;</code> &ndash; 
The minimum length of the peptides to place in the index. Default = 6.
</li>
	
<LI>
<code>max-length &lt;int&gt;</code> &ndash; 
The maximum length of the peptides to place in the index. Default = 50.
</li>

<LI>
<code>--enzyme trypsin|trypsin/p|chymotrypsin|elastase|clostripain|cyanogen-bromide|idosobenzoate|proline-endopeptidase|staph-protease|asp-n|lys-c|lys-n|arg-c|glu-c|pepsin-a|elastase-trypsin-chymotrypsin|no-enzyme</code> &ndash;
Enzyme to use for <i>in silico</i> digestion of protein
sequences. Used in conjunction with the <code>digestion</code>
and <code>missed-cleavages</code> options. Use 'no-enzyme' for
non-specific digestion.  Digestion rules are as follows: enzyme name
[cuts after one of these residues]|{but not before one of these
residues}. trypsin [RK]|{P}, trypsin/p [RK]|[], elastase [ALIV]|{P},
chymotrypsin [FWYL]|{P}, clostripain [R]|[], cyanogen-bromide [M]|[],
iodosobenzoate [W]|[], proline-endopeptidase [P]|[], staph-protease [E]|[],
elastase-trypsin-chymotrypsin [ALIVKRWFY]|{P}, asp-n []|[D], lys-c [K]|{P},
lys-n []|[K], arg-c [R]|{P}, glu-c [DE]|{P}, pepsin-a [FL]|{P}.
Default = trypsin.
</li>

<li>
<code>custom-enzyme &lt;residues before cleavage | residues after cleavage
&gt;</code> &ndash
Specify rules for <i>in silico</i> digestion of protein sequences.
Overrides the <code>enzyme</code> option.  Two lists of residues are
given enclosed in square brackets or curly braces and separated by a
|. The first list contains residues required/prohibited before the
cleavage site and the second list is residues after the cleavage site.
If the residues are required for digestion, they are in square
brackets, '[' and ']'.  If the residues prevent digestion, then they
are enclosed in curly braces, '{' and '}'.  Use X to indicate all
residues.  For example, trypsin cuts after R or K but not before P
which is represented as <code>[RK]|{P}</code>.  AspN cuts after any
residue but only before D which is represented as <code>[X]|[D]</code>.
</li>

<LI>
<code>digestion &lt;full-digest|partial-digest&gt;</code> &ndash;
Degree of digestion used to generate peptides . Either both ends
(full-digest) or at least one end (partial-digest) of a peptide must
conform to enzyme specificity rules. Used in conjunction with
the <code>enzyme</code> or <code>custom-enzyme</code> option
when <code>enzyme</code> is not set to to 'no-enzyme'.   Default
full-digest.
</li>

<LI>
<code>missed-cleavages &lt;T|F&gt;</code> &ndash;
Allow missed cleavage sites within a peptide.  When used with 
<code>enzyme</code> and set to true, includes peptides containing
one or more potential cleavage sites. Default = F.
</li>

<LI>
<code>isotopic-mass &lt;average|mono&gt;</code> &ndash; 
Specify the type of isotopic masses to use when calculating the
peptide mass. Default = average.
</li>
   
<li>
<code>&lt;A-Z&gt &lt;float&gt;</code> &ndash; 
Specify static modifications.  This is a mass change applied to the
given amino acid (in single-letter-code A thru Z) for every peptide in
which it occurs. Use the <code>mod</code> option for generating
peptides both with and without the mass change.  Default C=57.
</li>
</ul>

<HR></hr>

<P>
<A HREF="crux.html">Crux home</A></P>

</blockquote>
</BODY>
</HTML>
